app/registries/foreman/plugin/fact_parser_registry.rb

```diff
@@ -0,0 +1,44 @@
+module Foreman
+  class Plugin
+    class FactParserRegistry
+      def initialize
+        @parsers = {}.with_indifferent_access
+      end
+
+      # @return Parser for given key
+      # If the default parser is registered and unknown +key+ is used,
+      # the default parser is returned
+      #
+      # === Examples
+      #   FactParserRegistry[:puppet] # => PuppetParser
+      #   FactParserRegistry[:no_name] # => NoNameParser
+      #   FactParserRegistry[:unknown_key] # => PuppetParser (Puppet parser is registered as default)
+      def [](key)
+        @parsers[key]
+      end
+
+      # Registering a fact parser (at +parser+ variable) to Parser Registry for given +key+.
+      # It's possible to mark +parser+ as default option for unknown +key+.
+      # +key+ - key symbol to find parser, eg. :puppet
+      # +parser+ - class of parser, eg. PuppetParser
+      # +default+ - mark if the parser will be default option for unknown keys, false by default
+      #
+      # === Examples
+      #   FactParserRegistry.registry(:puppet, PuppetParser, true)
+      #   FactParserRegistry.registry(:no_name, NoNameParser)
+      def register(key, parser, default = false)
+        if (old_parser = @parsers[key]) && old_parser != @parsers.default
+          Rails.logger.warn("WARNING: Parser #{old_parser} for type #{key} is replaced with #{parser}")
+        end
+
+        @parsers.default = parser if default
+        @parsers[key.to_sym] = parser
+      end
+
+      # removing parser from registry
+      def unregister(key)
+        @parsers[key] = nil
```

Comment:
Should it delete? I wonder if otherwise the `default` may not work correctly.
```suggestion
        @parsers.delete(key)
```